tools/runners/moore.py

```diff
@@ -0,0 +1,49 @@
+import re
+from BaseRunner import BaseRunner
+
+
+class moore(BaseRunner):
+    def __init__(self):
+        super().__init__("moore", "moore")
+
+        self.url = "http://llhd.io/"
+
+    def prepare_run_cb(self, tmp_dir, params):
+        self.cmd = [self.executable]
+
+        for incdir in params['incdirs']:
+            self.cmd.append('-I')
+            self.cmd.append(incdir)
+
+        if params['top_module'] != '':
+            self.cmd.append('-e')
+            self.cmd.append(params['top_module'])
+        else:
+            regex = re.compile(r'module\s+(\w+)\s*[#(;]')
+            for fn in params['files']:
+                with open(fn) as f:
+                    m = regex.search(f.read())
+                    if m:
+                        self.cmd.append('-e')
+                        self.cmd.append(m.group(1))
```

Comment:
Is it necessary to specify list of modules to `moore`?
If it is then maybe this part should be moved to the main `BaseRunner` class as it looks like a fairly generic feature that other runners could also use to find the top module when it is not specified explicitly
@tgorochowik ?

Reply:
What exactly is this doing? Are you looking for an actual module name in the test file and returning the very first one you find? Is this necessary?

@piotr-binkowski are there any other runners that need this or could work better if they had this information? I agree that it seems generic, so maybe extracting it makes sense if it could be useful anywhere else.

Reply:
Something like this could for example eliminate [Verilator warnings](https://symbiflow.github.io/sv-tests/#Verilator|hdlconv|hdlconvertor_p234) about multiple top level modules.
But after rethinking this, it appears that this automated approach can create new issues when first module is not the top one

Reply:
So maybe we should make the top module tag required for files with more than one module in them, and optional only for files with a single/no module?

Reply:
> What exactly is this doing? Are you looking for an actual module name in the test file and returning the very first one you find? Is this necessary?

Indeed that's the idea. I have noticed that not all test provide a top-level module name, especially the "simple" unit tests lack that information. If no top-level module is specified, the regex just picks the first module in the file. Pretty hacky and unelegant, but gets the job done for most of the single-module tests.

Reply:
> Is it necessary to specify list of modules to `moore`?

No it's not needed. If no module is passed to `moore` for elaboration, the compiler will just parse the source files. However most of the interesting business (type checking, codegen) happens upon elaboration, when the type and parameter bindings have been fully established. I suspect this is the case for many other tools as well…

Actually coming to think of it, most tools actually require you to mention the modules you are interested in (simulators, synthesizers), even though some simulators might fall back to the *"modules never instantiated = top modules"* thing.

Maybe we should require test cases to always provide one or more top-level modules in general? Then this code could live in the test generator space and tools can always assume to be given the relevant module names.